File: web/src/test/java/me/robinvdb/runetracker/database/DataPointDatabaseStubTests.java
package me.robinvdb.runetracker.database;

import me.robinvdb.runetracker.domain.DataPoint;
import org.junit.Test;

import java.util.ArrayList;
import java.util.Collection;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatExceptionOfType;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

public class DataPointDatabaseStubTests {
    @Test
    public void testAddAddsADataPointToTheDatabaseIfItDoesNotYetExist() {
        DataPointDatabase database = new DataPointDatabaseStub();
        DataPoint point = mock(DataPoint.class);

        database.add(point);

        assertThat(database.all()).contains(point);
    }

    @Test
    public void testsAddThrowsExceptionWhenDataPointAlreadyExists() {
        DataPointDatabase database = new DataPointDatabaseStub();
        DataPoint point = mock(DataPoint.class);
        when(point.getId()).thenReturn(0);
        database.add(point);

        assertThatExceptionOfType(DatabaseException.class).isThrownBy(() -> database.add(point))
                .withMessage("DataPoint with id 0 already exists");
    }

    @Test
    public void testDeleteDeletesTheDataPoint() {
        DataPointDatabase database = new DataPointDatabaseStub();
        DataPoint point = mock(DataPoint.class);
        database.add(point);

        database.delete(point);

        assertThat(database.all()).doesNotContain(point);

    }

    @Test
    public void testGetSkillReturnsTheCorrectSkillObject() {
        DataPointDatabase database = new DataPointDatabaseStub();
        DataPoint point = mock(DataPoint.class);
        int id = 0;

        when(point.getId()).thenReturn(id);
        database.add(point);

        assertThat(database.get(id)).contains(point);
    }

    @Test
    public void testAddAllAddsThemAllToTheDatabase() {
        DataPointDatabase database = new DataPointDatabaseStub();

        Collection<DataPoint> dataPoints = new ArrayList<>();
        int iterations = 500;
        for (int i = 0; i < iterations; i++) {
            DataPoint dataPoint = mock(DataPoint.class);
            when(dataPoint.getId()).thenReturn(i);
            dataPoints.add(dataPoint);
        }
        database.addAll(dataPoints);

        assertThat(database.all()).containsAll(dataPoints);
    }
}

File: web/src/test/java/me/robinvdb/runetracker/database/DataPointDatabaseStubTests.kt
package me.robinvdb.runetracker.database

import me.robinvdb.runetracker.domain.DataPoint
import org.junit.Test

import java.util.ArrayList

import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatExceptionOfType
import org.mockito.Mockito.mock
import org.mockito.Mockito.`when`

class DataPointDatabaseStubTests {
    @Test
    fun testAddAddsADataPointToTheDatabaseIfItDoesNotYetExist() {
        val database = DataPointDatabaseStub()
        val point = mock(DataPoint::class.java)

        database.add(point)

        assertThat(database.all()).contains(point)
    }

    @Test
    fun testsAddThrowsExceptionWhenDataPointAlreadyExists() {
        val database = DataPointDatabaseStub()
        val point = mock(DataPoint::class.java)
        `when`<Int>(point.id).thenReturn(0)
        database.add(point)

        assertThatExceptionOfType(DatabaseException::class.java).isThrownBy { database.add(point) }
                .withMessage("DataPoint with id 0 already exists")
    }

    @Test
    fun testDeleteDeletesTheDataPoint() {
        val database = DataPointDatabaseStub()
        val point = mock(DataPoint::class.java)
        database.add(point)

        database.delete(point)

        assertThat(database.all()).doesNotContain(point)

    }

    @Test
    fun testGetSkillReturnsTheCorrectSkillObject() {
        val database = DataPointDatabaseStub()
        val point = mock(DataPoint::class.java)
        val id = 0

        `when`<Int>(point.id).thenReturn(id)
        database.add(point)

        assertThat(database[id]).contains(point)
    }

    @Test
    fun testAddAllAddsThemAllToTheDatabase() {
        val database = DataPointDatabaseStub()

        val dataPoints = ArrayList<DataPoint>()
        val iterations = 500
        for (i in 0 until iterations) {
            val dataPoint = mock(DataPoint::class.java)
            `when`<Int>(dataPoint.id).thenReturn(i)
            dataPoints.add(dataPoint)
        }
        database.addAll(dataPoints)

        assertThat(database.all()).containsAll(dataPoints)
    }
}
